BedrockServer.cpp

```diff
@@ -2209,20 +2035,268 @@ void BedrockServer::_finishPeerCommand(unique_ptr<BedrockCommand>& command) {
+                        // socket thread.
+                        lock_guard lock(m);
+                        cv.notify_all();
+                    };
+
+                    // Ok, none of above synchronization code gets called unless the command has a socket to respond
+                    // on.
+                    bool hasSocket = command->socket;
+                    if (hasSocket) {
+                        // Set the destructor callback for when the command finishes.
+                        command->destructionCallback = &callback;
+
+                        // And lock the mutex so the command can't complete until we are in `wait` below.
+                        lock.lock();
+                    }
+
+                    // Now we'll queue this command in one of three queues.
+                    auto _syncNodeCopy = atomic_load(&_syncNode);
+                    if (_syncNodeCopy && _syncNodeCopy->getState() == SQLiteNode::STANDINGDOWN) {
+                        _standDownQueue.push(move(command));
+                    } else {
+                        if (_version != _leaderVersion.load()) {
+                            SINFO("Immediately escalating " << command->request.methodLine << " to leader due to version mismatch.");
+                            _syncNodeQueuedCommands.push(move(command));
+                        } else {
+                            SINFO("Queuing new '" << command->request.methodLine << "' command from local client, with "
+                                  << _commandQueue.size() << " commands already queued.");
+                            _commandQueue.push(move(command));
+                        }
+                    }
+
+                    // Now that the command is queued, we wait for it to complete (if it's has a socket). When it's
+                    // destructionCallback fires, this will stop blocking and we can move on to the next request.
+                    if (hasSocket) {
+                        cv.wait(lock);
+                    }
+                }
+            } else {
+                // If we weren't able to deserialize a complete request, and we're shutting down, give up.
+                if (_shutdownState.load() != RUNNING && _lastChance && _lastChance < STimeNow()) {
+                    SINFO("Closing socket " << s->id << " with incomplete data and no pending command: shutting down.");
+                }
+            }
+        } else if (s->state.load() == STCPManager::Socket::SHUTTINGDOWN) {
+            // We do nothing in this state, we just wait until the next iteration of poll and let the CLOSED
+            // case run. This block just prevents default warning from firing.
+        } else {
+            SWARN("Socket in unhandled state: " << s->state);
+        }
+
+        // At the bottom of the loop, we need to `recv` again to either notice that we're disconnected, or get the next
+        // request.
```

Comment:
Are we missing something below this comment, like a call to `s->recv` or is this comment out of date?

Reply:
Read that comment like:

> We are at the bottom of the loop. We need to `recv` again to either notice that we're disconnected, or get the next request, so back to the top we go!